### ISSUE TITLE
Can't set constructor on document type in latest @types/mongoose

### ISSUE THREAD

**Issue by tomyam1** [DefinitelyTyped/DefinitelyTyped#37088]
The following code fails with TS2345 after 26587aa3aa75e0021180a9a7f9b4cb45e036002b:

Tested on Typescript 3.5.3.

```typescript
import * as Mongoose from 'mongoose'

export interface IPersonDocument extends Mongoose.Document {
    // The failure happens because we have constructor in IPersonDocument
    constructor: IPersonModel

    firstName: string
}

export interface IPersonModel extends Mongoose.Model<IPersonDocument> {

}

let Person: IPersonModel;

/*
Fails with typescript 3.5.3
TS2345: Argument of type '{ firstName: string; }' is not assignable to parameter of type 'Partial<IPersonDocument>'.   Types of property 'constructor' are incompatible.     Type 'Function' is missing the following properties from type 'IPersonModel': watch, translateAliases, bulkWrite, findById, and 56 more.
 */
const john = new Person({
    firstName: 'John',
});
```

@richarddd 
@armanio123

**Comment by richarddd** [comment]
Aha! The tests do not cover this usecase. We should probably handle that as well as a unary type! Could you create a PR?

**Comment by tomyam1** [comment]
`DeepPartial<T> & { constructor?: T}` gives the same error.
But ignoring the constructor solved it `Pick<DeepPartial<T>, Exclude<keyof DeepPartial<T>, 'constructor'>>`.

But then we got to other with this type checking. 

For instance, this valid code fails.

A string is valid source for an ObjectId but DeepPartial won't allow it.

I actually think it's best to remove the type checking from new function since DeepPartial is way too simplistic solution.

```typescript
import * as Mongoose from 'mongoose'

export interface IPersonDocument extends Mongoose.Document {
    father: Mongoose.Types.ObjectId // id of father record
}

export interface IPersonModel extends Mongoose.Model<IPersonDocument> {

}

let Person: IPersonModel;


const john = new Person({
    // TS2559: Type 'string' has no properties in common with type 'DeepPartial<ObjectId>'.
    father: 'valid-object-id'
});
```

**Comment by richarddd** [comment]
> `DeepPartial<T> & { constructor?: T}` gives the same error.
> But ignoring the constructor solved it `Pick<DeepPartial<T>, Exclude<keyof DeepPartial<T>, 'constructor'>>`.
> 
> But then we got to other with this type checking.
> 
> For instance, this valid code fails.
> 
> A string is valid source for an ObjectId but DeepPartial won't allow it.
> 
> I actually think it's best to remove the type checking from new function since DeepPartial is way too simplistic solution.
> 
> ```ts
> import * as Mongoose from 'mongoose'
> 
> export interface IPersonDocument extends Mongoose.Document {
>     father: Mongoose.Types.ObjectId // id of father record
> }
> 
> export interface IPersonModel extends Mongoose.Model<IPersonDocument> {
> 
> }
> 
> let Person: IPersonModel;
> 
> 
> const john = new Person({
>     // TS2559: Type 'string' has no properties in common with type 'DeepPartial<ObjectId>'.
>     father: 'valid-object-id'
> });
> ```

We can add checks for object id as well. Removing type checking is not a viable solution as it introduces other errors along the line :)

I think we should make a new PR which covers all these use cases and profit from better type checking in mongoose :+1:

**Comment by tomyam1** [comment]
This is just the tip of the iceberg...
What about strings as input for Date objects?
Or any other cast the mongoose makes.

It's very hard to cover all the use cases.

Maybe adding typechecking using a generic would be  a better option?

Something like:

`new<SrcType=DeepPartial<T>>(doc?: SrcType): T;`

**Comment by richarddd** [comment]
> This is just the tip of the iceberg...
> What about strings as input for Date objects?
> Or any other cast the mongoose makes.
> 
> It's very hard to cover all the use cases.
> 
> Maybe adding typechecking using a generic would be a better option?
> 
> Something like:
> 
> `new<SrcType=DeepPartial<T>>(doc?: SrcType): T;`

What casts does mongoose make?

We can solve this easily:

```typescriptt
 export type MongooseDeepPartial<T> = {
    [P in keyof T]?: T[P] extends (infer U)[]
      ? MongooseDeepPartial<U>[]
      : T[P] extends Date
      ? T[P] | string
      : T[P] extends mongoose.Types.ObjectId
      ? mongoose.Types.ObjectId | string
      : T[P] extends object
      ? MongooseDeepPartial<T[P]>
      : T[P]
  }
```

Also changed constructor like this adding the optional constructor:
`new (doc?: MongooseDeepPartial<T & { constructor?: T }>): T`

When we have added all the casts i will create a new PR and we'll solve this together @tomyam1  :)

**Comment by tomyam1** [comment]
@richarddd Sorry, this seems like an overly complicated solution and constant chance for all the valid use cases that this type-checking prevents.

Since we cannot use generic typings here (TS1092) I suggest that we revert back to `any`.

@armanio123 Any chance you weight in on this (since you approved the [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36517) that causes this issue)?

**Comment by richarddd** [comment]
> @richarddd Sorry, this seems like an overly complicated solution and constant chance for all the valid use cases that this type-checking prevents.
> 
> Since we cannot use generic typings here (TS1092) I suggest that we revert back to `any`.
> 
> @armanio123 Any chance you weight in on this (since you approved the [PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36517) that causes this issue)?

Reverting back to *any* would be the worst possible "solution" to the problem. The typings above solves all the mapping issues I could think of.

What's the point of using TS if we "solve" problems by falling back to *any* as soon as we run into tricky situations?

When changing to `DeepPartial` I immediately found bugs in the tests which passed fine when using *any*.

I will create a PR with this fix.

**Comment by tomyam1** [comment]
Your recent code still doesn't compile on our code base:

```
TS2345: 
...
  Types of property 'account' are incompatible.     Type 'string' is not assignable to type 'MongooseDeepPartial<ObjectId> | MongooseDeepPartial<IAccountDocument>'.
```

**Comment by tomyam1** [comment]
Fixed by c860c1dc7ba152a7e09b518eee900125a1a3bc3d
